src/Comparison.dfy

```diff
@@ -0,0 +1,93 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "Wrappers.dfy"
+include "Relations.dfy"
+module Comparison {
+
+  import opened Wrappers
+  import opened Relations
+
+  datatype Cmp = Lt | Eq | Gt {
+    function method Flip(): Cmp {
+      match this
+        case Lt => Gt
+        case Eq => Eq
+        case Gt => Lt
+    }
+
+    const Le? := this != Gt
+    const Ge? := this != Lt
+
+    static function method ComputeTransitivity(c0: Cmp, c1: Cmp): Option<Cmp> {
+      match (c0, c1)
+        case (Lt, Lt) => Some(Lt)
+        case (Lt, Eq) => Some(Lt)
+        case (Lt, Gt) => None
+        case (Eq, Lt) => Some(Lt)
+        case (Eq, Eq) => Some(Eq)
+        case (Eq, Gt) => Some(Gt)
+        case (Gt, Lt) => None
+        case (Gt, Eq) => Some(Gt)
+        case (Gt, Gt) => Some(Gt)
+    }
+  }
+   datatype Comparison<!T> = Comparison(cmp: (T, T) -> Cmp) {
+    function method Compare(t0: T, t1: T): Cmp {
+      cmp(t0, t1)
+    }
+
+     predicate Complete??(t0: T, t1: T) {
+      cmp(t0, t1) == cmp(t1, t0).Flip()
+    }
+
+    predicate Transitive??(t0: T, t1: T, t2: T) {
+      cmp(t0, t1).Le? && cmp(t1, t2).Le? ==> cmp(t0, t2).Le?
+    }
+
+    predicate Complete?(ts: set<T>) {
+      forall t0, t1 | t0 in ts && t1 in ts :: Complete??(t0, t1)
+    }
+
+    predicate Transitive?(ts: set<T>) {
+      forall t0, t1, t2 | t0 in ts && t1 in ts && t2 in ts :: Transitive??(t0, t1, t2)
+    }
+
+    predicate {:opaque} Valid?(ts: set<T>) {
+      Complete?(ts) && /* Antisymmetric?(ts) && */ Transitive?(ts)
```

Comment:
```suggestion
      Complete?(ts) && */ Transitive?(ts)
```

Just following the general rule of not checking in commented out code. We could put a comment saying something like "we want to include `Antisymmetric(ts)` as well but it isn't necessary and makes verification timeout" (for example, I don't actually know in this case).